[MSE] MediaSourcePrivate::buffered() shouldn't call its client

jyavenard opened the pull request:
#### bf1d50b780f12c0892b309371ce94423045002a8
<pre>
[MSE] MediaSourcePrivate::buffered() shouldn&apos;t call its client
<a href="https://bugs.webkit.org/show_bug.cgi?id=266064">https://bugs.webkit.org/show_bug.cgi?id=266064</a>
<a href="https://rdar.apple.com/119369641">rdar://119369641</a>

Reviewed by Youenn Fablet.

Have MediaSourcePrivate be the owner of the buffered range and have the MediaSource
query its MediaSourcePrivate instead.
By keeping it in the MediaSourcePrivate we can remove a synchronous call which will
no longer be possible once the MediaSourcePrivate runs in a different queue than the MediaSource.

Similarly, also make MediaSource not hold its duration, but have the MediaSourcePrivate do so.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isBuffered const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::duration const):
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::seekable):
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::monitorSourceBuffers):
(WebCore::MediaSource::setDurationInternal):
(WebCore::MediaSource::detachFromElement):
(WebCore::MediaSource::updateBufferedIfNeeded):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::duration const):
(WebCore::MediaSourcePrivate::bufferedChanged):
(WebCore::MediaSourcePrivate::buffered const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::bufferedChanged):
(WebKit::RemoteMediaSourceProxy::buffered const): Deleted. Implemented in the base class.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::durationChanged): ensure that the base class durationChanged is called.
(WebKit::MediaSourcePrivateRemote::bufferedChanged):
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/271794@main">https://commits.webkit.org/271794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71a444ca36838dd33d0a644e33c69d3f54e365c3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32156 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26826 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30268 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10501 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5587 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26830 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29926 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6966 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25298 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5913 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26383 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33496 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27062 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26806 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32266 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6004 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4198 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30041 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7743 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7046 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6548 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6528 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -184,7 +184,9 @@ MediaTime MediaSource::currentTime() const
-    return m_buffered;
+    if (isClosed())
+        return PlatformTimeRanges::emptyRanges();
+    return m_private->buffered();
```
One liner?

youennf commented on Source/WebCore/platform/graphics/MediaSourcePrivate.h:
```diff
@@ -76,7 +77,7 @@ class WEBCORE_EXPORT MediaSourcePrivate
     virtual void notifyActiveSourceBuffersChanged() = 0;
     virtual void durationChanged(const MediaTime&);
-    virtual void bufferedChanged(const PlatformTimeRanges&) { }
+    virtual void bufferedChanged(const PlatformTimeRanges& buffered); // Base class method must be called in overrides.
```
buffered name does not bring much.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1200,17 +1202,20 @@ void MediaSource::sourceBufferBufferedChanged()
+        if (buffered == m_private->buffered())
             return;
-        m_private->bufferedChanged(m_buffered);
+        m_private->bufferedChanged(buffered);
```
Should we think of WTFMoving buffered?

jyavenard replied:
it's a const reference for now. I'll take this as follow-up

youennf commented on Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
```diff
@@ -111,6 +111,7 @@ MediaSourcePrivate::AddStatus MediaSourcePrivateRemote::addSourceBuffer(const Co
 
 void MediaSourcePrivateRemote::durationChanged(const MediaTime& duration)
 {
+    MediaSourcePrivate::durationChanged(duration);
```
Maybe describe why this line is added (fly by fix?)

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -283,12 +285,12 @@ Ref<TimeRanges> MediaSource::seekable()
     // ↳ If duration equals NaN:
     // Return an empty TimeRanges object.
-    if (m_duration.isInvalid())
+    if (isClosed() || m_duration.isInvalid())
```
This is not in the spec.
If this is to handle the case of stopped sources, maybe we can set m_duration to invalid in stop() method.

jyavenard replied:
actually, it is kind in the spec, `If the [readyState](https://w3c.github.io/media-source/#dom-mediasource-readystate) attribute is "[closed](https://w3c.github.io/media-source/#dom-readystate-closed)" then return NaN and abort these steps.` so if closed duration is NaN

jyavenard replied:
I've reworked MediaSource::duration so that it gets queried via a method; this will make it duration() returns invalid if the media source is closed (which is per spec)

jyavenard committed:
[MSE] MediaSourcePrivate::buffered() shouldn't call its client
https://bugs.webkit.org/show_bug.cgi?id=266064
rdar://119369641

Reviewed by Youenn Fablet.

Have MediaSourcePrivate be the owner of the buffered range and have the MediaSource
query its MediaSourcePrivate instead.
By keeping it in the MediaSourcePrivate we can remove a synchronous call which will
no longer be possible once the MediaSourcePrivate runs in a different queue than the MediaSource.

Similarly, also make MediaSource not hold its duration, but have the MediaSourcePrivate do so.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isBuffered const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::duration const):
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::seekable):
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::monitorSourceBuffers):
(WebCore::MediaSource::setDurationInternal):
(WebCore::MediaSource::detachFromElement):
(WebCore::MediaSource::updateBufferedIfNeeded):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::duration const):
(WebCore::MediaSourcePrivate::bufferedChanged):
(WebCore::MediaSourcePrivate::buffered const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::bufferedChanged):
(WebKit::RemoteMediaSourceProxy::buffered const): Deleted. Implemented in the base class.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::durationChanged): ensure that the base class durationChanged is called.
(WebKit::MediaSourcePrivateRemote::bufferedChanged):
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:

Canonical link: https://commits.webkit.org/271794@main
